### PR TITLE
Allow to query the total supply from the contract runtime

### LINKED ISSUE #4221: add a way to fetch total supply with a contract runtime request
Add a `ContractRuntimeRequest` variant that would explicitly request the value of the total supply under a specific state root hash.

This can be achieved with a `Query` request already, but adding a dedicated variant will avoid leaking the details of the global state and it will make the code more readable.

### PR DESCRIPTION
**Pull request by real-felix**
Fix #4221

### CODE REVIEW

**Comment by EdHastingsCasperAssociation** [comment]
can we discuss this during standup?

File: execution_engine/src/core/engine_state/mod.rs [hunk 1]
```diff
@@ -785,6 +785,40 @@ where
         Ok(BalanceResult::Success { motes, proof })
     }
 
+    /// Get the total supply for an era targeted by the state hash.
```
Comment (fizyk20) [inline execution_engine/src/core/engine_state/mod.rs]: I wouldn't say "for an era" here. Execution engine is ignorant of eras. It just deals with the global state as it is at some particular hash. So, maybe something like:

> Get the total supply of tokens at the given state hash.

File: node/src/effect.rs [hunk 2]
```diff
@@ -2017,6 +2017,28 @@ impl<REv> EffectBuilder<REv> {
         .await
     }
 
+    /// Returns a map of validators weights for all eras as known from `root_hash`.
+    ///
+    /// This operation is read only.
+    #[allow(unused)] //TODO remove in the next ticket implementation.
+    pub(crate) async fn get_era_total_supply(
```
Comment (fizyk20) [inline node/src/effect.rs]: Again, I wouldn't mention eras here. The fact that total supply is constant within an era is just an implementation detail.

Just something like `get_total_supply` is fine.

File: node/src/effect.rs [hunk 3]
```diff
@@ -2017,6 +2017,28 @@ impl<REv> EffectBuilder<REv> {
         .await
     }
 
+    /// Returns a map of validators weights for all eras as known from `root_hash`.
+    ///
+    /// This operation is read only.
```
Comment (fizyk20) [inline node/src/effect.rs]: Copy-pasted comment?

**Comment by fizyk20** [comment]
Per the discussion at stand-up, let's not change the execution engine itself, but rather have the code in the contract runtime component servicing the `GetTotalSupply` request call the `query` method on the execution engine.

File: execution_engine/src/core/engine_state/mod.rs (hunk 1 above)
Comment (EdHastingsCasperAssociation) [reply execution_engine/src/core/engine_state/mod.rs]: bart is correct

File: node/src/effect.rs (hunk 2 above)
Comment (real-felix) [reply node/src/effect.rs]: Done

File: node/src/effect.rs (hunk 3 above)
Comment (real-felix) [reply node/src/effect.rs]: Done

**Review by fizyk20 — APPROVED** [review]

**Review by AlexanderLimonov — APPROVED** [review]

**Review by EdHastingsCasperAssociation — APPROVED** [review]

**Review by mpapierski — APPROVED** [review]

**Pull request merged**